From 09b2b1b57a7a7d79b9cebf7d0b0c28091b8dea68 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 30 Mar 2022 10:57:24 +0200
Subject: [PATCH 06/19] user32: Introduce SetActiveWindow user driver callback.

CW-Bug-Id: #19553
CW-Bug-Id: #19644
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=2155

  87.7% dlls/win32u/
  12.2% include/wine/
diff --git a/dlls/win32u/driver.c b/dlls/win32u/driver.c
index 2937f0d7248..8c128b96607 100644
--- a/dlls/win32u/driver.c
+++ b/dlls/win32u/driver.c
@@ -873,6 +873,10 @@ static BOOL nulldrv_ScrollDC( HDC hdc, INT dx, INT dy, HRGN update )
                         hdc, rect.left - dx, rect.top - dy, SRCCOPY, 0, 0 );
 }
 
+static void nulldrv_SetActiveWindow( HWND hwnd, HWND prev )
+{
+}
+
 static void nulldrv_SetCapture( HWND hwnd, UINT flags )
 {
 }
@@ -1325,6 +1329,7 @@ static const struct user_driver_funcs lazy_load_driver =
     nulldrv_ProcessEvents,
     nulldrv_ReleaseDC,
     nulldrv_ScrollDC,
+    nulldrv_SetActiveWindow,
     nulldrv_SetCapture,
     loaderdrv_SetDesktopWindow,
     nulldrv_SetFocus,
@@ -1412,6 +1417,7 @@ void __wine_set_user_driver( const struct user_driver_funcs *funcs, UINT version
     SET_USER_FUNC(ProcessEvents);
     SET_USER_FUNC(ReleaseDC);
     SET_USER_FUNC(ScrollDC);
+    SET_USER_FUNC(SetActiveWindow);
     SET_USER_FUNC(SetCapture);
     SET_USER_FUNC(SetDesktopWindow);
     SET_USER_FUNC(SetFocus);
diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 173f7668a11..438130704e9 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -1870,6 +1870,7 @@ static BOOL set_active_window( HWND hwnd, HWND *prev, BOOL mouse, BOOL focus )
 
     if (previous == hwnd)
     {
+        if (hwnd == NtUserGetForegroundWindow()) user_driver->pSetActiveWindow( hwnd, previous );
         if (prev) *prev = hwnd;
         goto done;
     }
@@ -1898,6 +1899,7 @@ static BOOL set_active_window( HWND hwnd, HWND *prev, BOOL mouse, BOOL focus )
             previous = wine_server_ptr_handle( reply->previous );
     }
     SERVER_END_REQ;
     if (!ret) return FALSE;
+    if (hwnd == NtUserGetForegroundWindow()) user_driver->pSetActiveWindow( hwnd, previous );
     if (prev) *prev = previous;
     if (previous == hwnd) goto done;
 
diff --git a/include/wine/gdi_driver.h b/include/wine/gdi_driver.h
index 26562bfef2b..9a93ad9324c 100644
--- a/include/wine/gdi_driver.h
+++ b/include/wine/gdi_driver.h
@@ -332,6 +332,7 @@ struct user_driver_funcs
     BOOL    (*pProcessEvents)(DWORD);
     void    (*pReleaseDC)(HWND,HDC);
     BOOL    (*pScrollDC)(HDC,INT,INT,HRGN);
+    void    (*pSetActiveWindow)(HWND, HWND);
     void    (*pSetCapture)(HWND,UINT);
     void    (*pSetDesktopWindow)(HWND);
     void    (*pSetFocus)(HWND);
-- 
2.43.0

